### DISCUSSION TITLE
Deploying terrakube with self hosted with docker-compose on single node

### DISCUSSION THREAD

**Discussion by thatsk** [terrakube-io/terrakube#434] [Q&A]
Hello Team,

I want to deploy terrakube with private Ip or public IP of a node with docker-compose installation but documentation is restricted to localhost only. So if someone guides me where to edit all the files it would be benefit for me to run in production without k8s

**Comment by thatsk** [comment]
one more observation is that if we want to move local docker volumes to relative paths of system that would be great.

**Comment by thatsk** [comment]
so that we can redeploy with consistent data anywhere

**Comment by thatsk** [comment]
can we replace all names from all .env files with node IP ?

**Comment by alfespa17** [comment]
Hello @thatsk 

You should be able to deploy terrakube using docker compose or just plain docker but it will require a couple of changes, for example:

- The datase is using H2 (in-memory database), you will have to add for example a postgresql so you can have persistency
- I think you could change the names for the node IP (maybe you would have to do some troubleshooting but It could work in theory, not sure if you will have some CORS issues with the UI)
-  You could also edit the DNS record in your node adding the values in your /etc/hosts

If you want to run a single node maybe you could just install minikube, it will be easier to setup.

https://docs.terrakube.io/getting-started/deployment/minikube 

Keep in mind that if you want to self hosted and use all the feature like the terraform registry, terraform login, run terraform remote state  operation you will have to use a valid TLS certificate to expose the API and Registry so terrafom CLI commands can work propertly( this is a restriction from terraform cli not Terrakube)

**Reply by thatsk** [reply]
My company don't want to introduce k8s. We have nomad based workload. We still use docker-compose. And you mean like just edit host file of machine with IP I thought I need to change in all env files is it true only updating etc hosts file like 
nodeiP terrakube-api
nodeiP terrakube-dex

**Reply by thatsk** [reply]
But without k8s also we are able to to deploy full fledged terrakube isn't it I really want this product to be deployed as other product don't have the feaute like this one

**Reply by alfespa17** [reply]
You should be able to deploy a fully working terrakube without using kubernetes (we do this in our test environment in GITPOD because we run it directly in the node without any docker container or even locally for testing purposes)

it will require some effort but it will work :)

Regarding the DNS records.

I think if should work if you if you just add something like "nodeiP terrakube-api" in your DNS records, you need to keep in mind that the api, registry and executor should be able to resolve the private or public ip so they can communicate.

**Reply by alfespa17** [reply]
I will share an example of how you could run Terrakube using docker compose with a persistent database maybe that can help you a little bit

**Reply by alfespa17** [reply]
Added PostgreSQL in docker-compose so you can have a persistent database https://github.com/AzBuilder/terrakube/pull/437.

I hope this can help you.

**Comment by thatsk** [comment]
Thanks just taken in mind if you have private up instance with already dns has it. And it has docker-compose so that once you start we should just access IP or associated dns will work . /Etc/hosts entry will work if you are using local but if you are in VPN network and your instance is in that network developer will not edit hosts entry they willaccess hostname

**Reply by alfespa17** [reply]
I guess that if you are using a VPN your company should have a private DNS where you can add the terrakube componente IPs so it can be resolve as any other internal application

**Comment by thatsk** [comment]
That's totally correct we have private dns  . And as we are saying we have single node terrakube with docker compose iam trying to install one more thing minio with local directory is also not working might be related to permission or user issue

**Comment by KieranP** [comment]
Hello @alfespa17 , just chiming in here. Tried to setup Terrakube in the same way (running with docker compose on an EC2 instances on a private subnet, accessible through VPN) by following https://github.com/AzBuilder/terrakube/blob/main/docker-compose/README.md

Got it started and could access the login page via the private IP, but got numerous console errors because by default, the web UI tried to access http://terrakube-dex and http://terrakube-api rather than just hosing request.hostname.

I tried to edit env-config.js and change the URLs to use the private IP address, and then down/up the docker containers again, but that wasn't enough. I tried editing the .env files also, but this caused more issues. In the end, I just added the /etc/host lines to my local machine as well (not ideal as I'll need to tell other engineers to do this, was hoping to avoid that).

After all that, I was able to access it via (http://terrakube-ui:3000) (with /etc/hosts pointing to private IP) and login but now I'm having some issues setting up a workspace. I was able to create an organisation ok, but it seems the default user doesn't have the right permissions to create the workspace.

We're a small company. Using K8s/Helm is overkill for us, since we only need one organisation with one workspace. It would be very helpful if the docker compose route were amended to:

1) Not require /etc/hosts on the local machine - could do this by making Dex/API requests to the same hostname as the current web UI request (i.e. if I access it via http://10.0.0.20:3000, then make Dex requests to http://10.0.0.20:5556 and API to http://10.0.0.20:8080) so that it doesn't require hardcoding domains
2) Instead of needing to update /etc/hosts on the host machine, utilize docker network to setup those hostnames so the containers can communicate with one another if needed (see https://docs.docker.com/engine/network/). This will make any change to /etc/hosts redundant.
3) Give the default user full admin access, which it does not seem to have currently, and maybe change it's name from "Lester" to "Administrator" or something ^_^

These three tweaks would make setup far simpler for people like myself who want to spin up a quick low resource Terrakube environment.

**Reply by alfespa17** [reply]
Hello @KieranP maybe you can check this gist, if explain a better way to use docker compose with traefik. 

https://gist.github.com/SolomonHD/b55be40146b7a53b8f26fe244f5be52e
